### PR TITLE
Fix typing with optional fields and open records

### PR DESCRIPTION
**Pull request by MaryamZi**
## Purpose
$title.

Fixes https://github.com/ballerina-platform/ballerina-lang/issues/40928

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/resources/test-src/record/record_assignability.bal [hunk 1]
```diff
@@ -0,0 +1,116 @@
+// Copyright (c) 2023 WSO2 LLC. (http://www.wso2.org) All Rights Reserved.
```
Comment (gimantha) [inline tests/jballerina-unit-test/src/test/resources/test-src/record/record_assignability.bal:1]:
```suggestion
// Copyright (c) 2023 WSO2 LLC. (http://www.wso2.org)
```

Comment (lochana-chathura) [inline tests/jballerina-unit-test/src/test/resources/test-src/record/record_assignability.bal:1]:
```suggestion
// Copyright (c) 2023 WSO2 LLC. (http://www.wso2.com).
```

File: tests/jballerina-unit-test/src/test/resources/test-src/types/never/never-type-negative.bal [hunk 2]
```diff
@@ -259,6 +259,9 @@ function testNeverRestFieldType() {
 
     record {||} a3 = {};
     record {|int x;never...; |} copy4 = a3;
+
+    record {|never?...; |} a4 = {};
```
Comment (lochana-chathura) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/never/never-type-negative.bal:263]:
```suggestion
    record {|never?...;|} a4 = {};
```

File: tests/jballerina-unit-test/src/test/resources/test-src/record/record_assignability_negative.bal [hunk 3]
```diff
@@ -41,3 +41,31 @@ function testClosedRecordAssignabilityNegative() {
...
+type R3 record {|
+    string y?;
+    int...;
+|};
+
+function testOpenRecordToRecordWithIncompatibleOptionalFieldTyping() {
+    R2 r1 = {x: 1, "y": 10};
+    R1 _ = r1;
+
+    record {| int|string...; |} r2 = {};
```
Comment (lochana-chathura) [inline tests/jballerina-unit-test/src/test/resources/test-src/record/record_assignability_negative.bal:62]:
```suggestion
    record {|int|string...;|} r2 = {};
```
Here and below

File: tests/jballerina-unit-test/src/test/resources/test-src/record/record_assignability.bal [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+type R11 record {|
+    readonly int? b;
+    int...;
+|};
+
+function testRecordToRecordWithOptionalFieldTypingRuntimePositive() {
+    R9 a = {};
+    R8 _ = a; // OK
+    assertTrue(<any> a is R8);
```
Comment (lochana-chathura) [inline tests/jballerina-unit-test/src/test/resources/test-src/record/record_assignability.bal:88]:
```suggestion
    assertTrue(<any>a is R8);
```

File: tests/jballerina-unit-test/src/test/resources/test-src/record/record_assignability_negative.bal (hunk 3 above)
Comment (MaryamZi) [reply tests/jballerina-unit-test/src/test/resources/test-src/record/record_assignability_negative.bal:62]: Same as https://github.com/ballerina-platform/ballerina-lang/pull/40074/files#r1271727926.

Comment (MaryamZi) [reply tests/jballerina-unit-test/src/test/resources/test-src/record/record_assignability_negative.bal:62]: Will change for consistency here though.

**Comment by MaryamZi** [comment]
Merging to a feature branch to test OpenAPI generation.

**Pull request merged**